### PR TITLE
Implement biomes

### LINKED ISSUE #326: Implement biomes (1.18)

### PR DESCRIPTION
**Pull request by JustTalDevelops**
Closes #326.

### CODE REVIEW

**Review by Lucaskyy — APPROVED** [review]
Looks good!

File: server/world/biome/biome.go [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+// Biome is a region in a world with distinct geographical features, flora, temperatures, humidity ratings,
+// and sky, water, grass and foliage colors.
+type Biome interface {
+	// Temperature returns the temperature of the biome.
+	Temperature() float64
+	// Rainfall returns the rainfall of the biome.
+	Rainfall() float64
+	// String returns the biome name as a string.
+	String() string
```
Comment (Sandertv) [inline server/world/biome/biome.go:11]: Do we strictly need this?

**Review by Sandertv — COMMENTED** [review]
This PR doesn't currently handle the sending of the biome definition list packet, but I think we should do so in Dragonfly as well, especially when we allow custom biomes.

**Comment by Lucaskyy** [comment]
Shouldn't these TODO's be implemented with this PR?

https://github.com/df-mc/dragonfly/blob/c07cb365801a61249bf4c30c0579c0517e27b066/server/world/world.go#L706-L707

https://github.com/df-mc/dragonfly/blob/master/server/world/world.go#L719-L720

File: server/world/biome/savanna_plateau_mutated.go [hunk 2]
```diff
@@ -0,0 +1,24 @@
+package biome
+
+// ShatteredSavannaPlateau ...
+type ShatteredSavannaPlateau struct{}
```
[the remark this answers is not in the document]
Comment (JustTalDevelops) [reply server/world/biome/savanna_plateau_mutated.go]: Updated.

File: server/world/generator/flat.go [hunk 3]
```diff
@@ -23,11 +26,14 @@ func (f Flat) GenerateChunk(_ world.ChunkPos, chunk *chunk.Chunk) {
...
+	min := int16(chunk.Range().Min())
 	for x := uint8(0); x < 16; x++ {
 		for z := uint8(0); z < 16; z++ {
-			for y := int16(0); y < l; y++ {
-				chunk.SetBlock(x, min+y, z, 0, m[l-y-1])
+			for y := int16(0); y < int16(chunk.Range().Max()); y++ {
+				if y < l {
+					chunk.SetBlock(x, min+y, z, 0, m[l-y-1])
+				}
+				chunk.SetBiome(x, min+y, z, uint32(f.Biome.EncodeBiome()))
```
[the remark this answers is not in the document]
Comment (JustTalDevelops) [reply server/world/generator/flat.go]: Changed.

File: server/world/biome/biome.go [hunk 4]
```diff
@@ -0,0 +1,33 @@
+package biome
```
[the remark this answers is not in the document]
Comment (JustTalDevelops) [reply server/world/biome/biome.go]: Moved.

File: server/world/biome/biome.go (hunk 1 above)
Comment (JustTalDevelops) [reply server/world/biome/biome.go]: No, but it can be useful in some cases. Would you like me to remove it?

**Comment by JustTalDevelops** [comment]
> Shouldn't these TODO's be implemented with this PR?
> 
> https://github.com/df-mc/dragonfly/blob/c07cb365801a61249bf4c30c0579c0517e27b066/server/world/world.go#L706-L707
> 
> https://github.com/df-mc/dragonfly/blob/c07cb365801a61249bf4c30c0579c0517e27b066/server/world/world.go#L719-L720

Updated.

**Comment by JustTalDevelops** [comment]
Could I have another review on this?

File: server/server.go [hunk 5]
```diff
@@ -622,6 +624,44 @@ func (server *Server) itemEntries() (entries []protocol.ItemEntry) {
 	return
 }
 
+// ashyBiome represents a biome that has any form of hash.
```
Comment (DaPigGuy) [inline server/server.go:627]:
```suggestion
// ashyBiome represents a biome that has any form of ash.
```

File: server/server.go [hunk 6]
```diff
@@ -622,6 +624,44 @@ func (server *Server) itemEntries() (entries []protocol.ItemEntry) {
 	return
 }
 
+// ashyBiome represents a biome that has any form of ash.
+type ashyBiome interface {
+	// Ash returns the ash of the biome.
+	Ash() float64
+	// WhiteAsh returns the white ash of the biome.
+	WhiteAsh() float64
```
Comment (Sandertv) [inline server/server.go:632]: Could maybe merge this into a single `Ash() (float64, float64)`.

File: server/server.go [hunk 7]
```diff
@@ -622,6 +624,44 @@ func (server *Server) itemEntries() (entries []protocol.ItemEntry) {
...
+}
+
+// sporingBiome represents a biome that has blue or red spores.
+type sporingBiome interface {
+	// BlueSpores returns the blue spores of the biome.
+	BlueSpores() float64
+	// RedSpores returns the red spores of the biome.
+	RedSpores() float64
```
[the remark this answers is not in the document]
Comment (JustTalDevelops) [reply server/server.go]: Merged.

File: server/server.go (hunk 6 above)
Comment (JustTalDevelops) [reply server/server.go]: Merged.

**Comment by JustTalDevelops** [comment]
Is there anything else you'd like me to change (aside from the stuff in the gophertunnel PR?)

File: server/server.go [hunk 8]
```diff
@@ -622,6 +624,42 @@ func (server *Server) itemEntries() (entries []protocol.ItemEntry) {
...
+	definitions := make(map[string]interface{})
+	for _, b := range world.Biomes() {
+		definition := map[string]interface{}{
+			"temperature": float32(b.Temperature()),
+			"downfall":    float32(b.Rainfall()),
+		}
+		if a, ok := b.(ashyBiome); ok {
+			ash, whiteAsh := a.Ash()
+			definition["ash"] = float32(ash)
+			definition["white_ash"] = float32(whiteAsh)
```
Comment (Sandertv) [inline server/server.go:651]: Can merge these 3 lines into a single line.

File: server/server.go [hunk 9]
```diff
@@ -622,6 +624,42 @@ func (server *Server) itemEntries() (entries []protocol.ItemEntry) {
...
+		}
+		if s, ok := b.(sporingBiome); ok {
+			blueSpores, redSpores := s.Spores()
+			definition["blue_spores"] = float32(blueSpores)
+			definition["red_spores"] = float32(redSpores)
```
Comment (Sandertv) [inline server/server.go:656]: Same as above.

**Review by Sandertv — COMMENTED** [review]
2 small last nitpicks, this looks good otherwise. Biomes are not currently resent when the biome is changed, but that is not a very big deal for now. We can do so at a later time.

**Comment by JustTalDevelops** [comment]
Anything else I can change?

**Comment by Sandertv** [comment]
Should be good to go I think!

**Comment by Sandertv** [comment]
Actually could you run `go mod tidy` once?

**Comment by JustTalDevelops** [comment]
There we go.

**Review by Sandertv — APPROVED** [review]
Looks good to me!

**Pull request merged**